aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessorTest.java

```diff
@@ -0,0 +1,227 @@
+
+        // Was the data merged?
+        assertTrue(latch.await(OPERATION_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private static <T extends Model> void arrangeDataEmittingSubscription(
+            AppSync appSync,
+            Class<T> clazz,
+            SubscriptionType subscriptionType,
+            GraphQLResponse<ModelWithMetadata<T>> response) throws DataStoreException {
+        Answer<Cancelable> answer = invocation -> {
+            final int startConsumerIndex = 1;
+            Consumer<String> onStart = invocation.getArgument(startConsumerIndex);
+            onStart.accept(RandomString.string());
+
+            final int dataConsumerIndex = 2;
+            Consumer<GraphQLResponse<ModelWithMetadata<T>>> onData = invocation.getArgument(dataConsumerIndex);
+            onData.accept(response);
+
+            return new NoOpCancelable();
+        };
+        arrangeSubscription(appSync, answer, clazz, subscriptionType);
+    }
+
+    private static void arrangeStartedSubscriptions(
+        AppSync appSync, List<Class<? extends Model>> classes, SubscriptionType[] subscriptionTypes) {
+        Answer<Cancelable> answer = invocation -> {
+            final int startConsumerIndex = 1;
+            Consumer<String> onStart = invocation.getArgument(startConsumerIndex);
+            onStart.accept(RandomString.string());
+            return new NoOpCancelable();
+        };
+        arrangeSubscriptions(appSync, answer, classes, subscriptionTypes);
+    }
+
+    private static void arrangeSubscriptions(
+            AppSync appSync,
+            Answer<Cancelable> answer,
+            List<Class<? extends Model>> classes,
+            SubscriptionType[] subscriptionTypes) {
+        Observable.fromIterable(classes)
+            .flatMap(modelClass -> Observable.fromArray(subscriptionTypes)
+                .map(subscriptionType -> Pair.create(modelClass, subscriptionType)))
+            .blockingForEach(pair -> arrangeSubscription(appSync, answer, pair.first, pair.second));
+    }
+
+    private static <T extends Model> void arrangeSubscription(
+            AppSync appSync, Answer<Cancelable> answer, Class<T> clazz, SubscriptionType subscriptionType)
+            throws DataStoreException {
+        AppSync stub = doAnswer(answer).when(appSync);
+        SubscriptionProcessor.SubscriptionMethod method =
+            SubscriptionProcessor.subscriptionMethodFor(stub, subscriptionType);
+        method.subscribe(eq(clazz), anyConsumer(), anyConsumer(), anyConsumer(), anyAction());
+    }
+
+    private static void verifySubscriptions(
+            AppSync appSync, List<Class<? extends Model>> classes, SubscriptionType[] subscriptionTypes) {
+        Observable.fromIterable(classes)
+            .flatMap(modelClass -> Observable.fromArray(subscriptionTypes)
+                .map(subscriptionType -> Pair.create(modelClass, subscriptionType)))
```

Comment:
Java's missing tuple :) I like it. I'll have to remember it's under android.util